yt/frontends/stream/data_structures.py

```diff
@@ -560,10 +560,14 @@ def __init__(
             axis_order=axis_order,
         )
         fields = list(stream_handler.fields["stream_file"].keys())
-        # This is the current method of detecting SPH data.
-        # This should be made more flexible in the future.
-        if ("io", "density") in fields and ("io", "smoothing_length") in fields:
-            self._sph_ptypes = ("io",)
+        sph_ptypes = []
+        for ptype in self.particle_types:
+            if (ptype, "density") in fields and (ptype, "smoothing_length") in fields:
+                sph_ptypes.append(ptype)
+        if len(sph_ptypes) == 1:
+            self._sph_ptypes = tuple(sph_ptypes)
+        elif len(sph_ptypes) > 1:
+            raise ValueError("Multiple SPH particle types are currently not supported!")
```

Comment:
It's probably unlikely, but I'm worried this might break existing applications. A warning + backward compatible behaviour would be more suited.

Reply:
It's never been possible for anyone to have multiple SPH `ptypes`, so there should not be anything broken. We should raise an error here because a lot of things in yt assume one SPH type.

Reply:
it's never been supported, yes, but the existing logic effectively ignores other sph types if any; so this error would break existing applications that have them. I guess such applications would actually be silently broken and it's good that this makes their bug apparent ?

Reply:
@neutrinoceros the point is that the code in `main` only recognizes one SPH type, and that is `io`, whereas this supports other type names while still only supporting one. If users are attempting to use multiple SPH ptypes, it is currently failing silently until they try to do something which requires that type to be recognized. So this is in fact an improvement because it's actually raising a specific error telling them what is wrong.

Reply:
ok, so that's what I was suspecting, and in that case I agree that it is an improvement. Thanks for clarifying !